inbound_test.go

```diff
@@ -142,3 +142,38 @@ func TestInboundConnection_CallOptions(t *testing.T) {
 		require.NoError(t, err, "Call through proxy failed")
 	})
 }
+
+func TestBlackhole(t *testing.T) {
+	ctx, cancel := NewContext(testutils.Timeout(time.Hour))
+
+	testutils.WithTestServer(t, nil, func(server *testutils.TestServer) {
+		serviceName := server.ServiceName()
+		handlerName := "test-handler"
+
+		server.Register(HandlerFunc(func(ctx context.Context, inbound *InboundCall) {
+			// cancel client context in handler so the client can return after being blackholed
+			defer cancel()
+
+			c, _ := InboundConnection(inbound)
+			require.NotNil(t, c)
+
+			state := c.IntrospectState(&IntrospectionOptions{})
+			require.Equal(t, 1, state.InboundExchange.Count, "expected exactly one inbound exchange")
+
+			// blackhole request
+			inbound.Response().Blackhole()
+
+			// give time for exchange to cleanup
+			time.Sleep(10 * time.Millisecond)
```

Comment:
instead of sleep, take a look at `testutils.WaitFor` -- it's a lot less flaky for testing since it'll sleep a small amount initially, check whether what we're waiting for is true, if not, sleep a little longer, and repeat.

The issue with a single `time.Sleep` is that it could finish in 1ms, in which case you're wasting 9ms, or it could take 11ms, in which case the test becomes flaky and fails. The only way to avoid flakiness is to use a huge value like a second or more, but then you're likely wasting 900ms for each run just doing nothing.

`WaitFor` is a compromise that lets you sleep (since some things can't be tested without waiting for some background work to be scheduled), without making the test slower or flaky.